app/models/specific_tube_rack_creation.rb

```diff
@@ -0,0 +1,436 @@
+  #
+  # This method creates a new PolyMetadatum record with the provided metadata key and tube rack barcode.
+  # The metadatable_type and metadatable_id are set to the class and ID of the tube rack, respectively.
+  # If the metadata record fails to save, an error is raised.
+  #
+  # @param [String] metadata_key The key for the metadata.
+  # @param [String] tube_rack_barcode The barcode of the tube rack to be used as the metadata value.
+  # @param [TubeRack] tube_rack The tube rack object to which the metadata will be added.
+  #
+  # @raise [StandardError] if the metadata record fails to save.
+  #
+  # @return [void]
+  def add_tube_rack_metadata(metadata_key, tube_rack_barcode, tube_rack)
+    pm =
+      PolyMetadatum.new(
+        key: metadata_key,
+        value: tube_rack_barcode,
+        metadatable_type: tube_rack.class,
+        metadatable_id: tube_rack.id
+      )
+    return if pm.save
+
+    raise StandardError, "New metadata for tube rack (key: #{metadata_key}, value: #{tube_rack_barcode}) did not save"
+  end
+
+  # Creates racked tubes based on the provided attributes and associates them with a new tube rack.
+  #
+  # This method iterates over the array of racked tube attributes provided in the rack_attributes
+  # hash.
+  # For each set of tube attributes, it calls the create_racked_tube method to create and associate
+  # the tube with the new tube rack.
+  #
+  # @param [Array<Hash>] racked_tubes An array of hashes, each containing attributes for a racked tube.
+  # @param [TubeRack] new_tube_rack The new tube rack object to which the tubes will be associated.
+  #
+  # @raise [StandardError] if any of the tube creation processes fail.
+  #
+  # @return [void]
+  def create_racked_tubes(racked_tubes, new_tube_rack)
+    # iterate through the rack attributes to create each tube
+    racked_tubes.each { |tube_attributes| create_racked_tube(tube_attributes, new_tube_rack) }
+  end
+
+  def create_racked_tube(tube_attributes, new_tube_rack)
+    tube = create_tube(tube_attributes)
+    link_tube_to_rack(tube, new_tube_rack, tube_attributes[:tube_position])
+  end
+
+  # Ensures that the provided tube barcode is unique.
+  #
+  # This method checks if a tube barcode already exists in the database.
+  # If the barcode is found, it raises a StandardError indicating that the barcode is already in use.
+  # If the barcode is not found, the method simply returns, allowing the process to continue.
+  #
+  # @param [String] tube_barcode The barcode of the tube to be checked for uniqueness.
+  #
+  # @raise [StandardError] if the tube barcode is already in use.
+  #
+  # @return [void]
+  def ensure_unique_tube_barcode(tube_barcode)
+    existing_tube_barcode_record = Barcode.includes(:asset).find_by(asset_id: tube_barcode)
+    return if existing_tube_barcode_record.nil?
+
+    error_message = "The tube barcode '#{tube_barcode}' is already in use, cannot continue."
+    raise StandardError, error_message
```

Comment:
Is this replicating behaviour that would happen already if you tried to insert a tube with a pre-existing barcode?

Reply:
Might be worth having a discussion about error handling for API endpoints in general, as I'm not very clear on what the best way to do all this is. This endpoint has to make multiple tube racks, each with multiple tubes. If anything goes wrong at any point I want the whole lot rolled back and a suitable human friendly error message passed back through the api response to be displayed to the user.
Does raising an explicit StandardError like this trigger the full rollback? And does that error message make it back to Limber and get displayed? Would I get the same result attempting an insert and that failing with a uniqueness error (given barcode is a different table than labwares / receptacles)? Does it still rollback everything and return a clear message to Limber?

Reply:
Shall we put it on our list to ask Stuart about?